lib/strategy/FragmentationStrategyShape.ts

```diff
@@ -0,0 +1,292 @@
+   * Evaluate if shape index information should be generated based on which quad has been handled and
+   * the shape config file provides. Generate the right iri for the shape indexes.
+   * The generation of the triples and the quad sink pushing is handled by other functions called by
+   * handleQuad.
+   * @param {RDF.Quad} quad - a quad
+   * @param {IQuadSink} quadSink - a quad sink
+   */
+  protected async handleQuad(quad: RDF.Quad, quadSink: IQuadSink): Promise<void> {
+    const iri = FragmentationStrategySubject.generateIri(quad, this.relativePath);
+    // If iri already has been handled, we do nothing
+    if (!this.irisHandled.has(iri)) {
+      for (const [ resourceIndex, { shape, directory }] of this.shapeMap) {
+        // Find the position of the first character of the container
+        const positionContainerResourceNotInRoot = iri.indexOf(`/${directory}/`);
+        const positionContainerResourceInRoot = iri.indexOf(`/${resourceIndex}`);
+        const positionContainer = positionContainerResourceNotInRoot !== -1 ?
+          positionContainerResourceNotInRoot :
+          positionContainerResourceInRoot;
+
+        // We use as an id the path of the iri until the resource identifier.
+        // It is a different discrimination mechanism than the one related to the iri because even
+        // if we don't want to add shape index information we might want to add shapeTreeLocator in
+        // every resource bounded by a shape.
+        const resourceId = `${iri.slice(0, Math.max(0, positionContainer))}/${resourceIndex}`;
+        if (positionContainer !== -1) {
+          const generate = this.shouldGenerate();
+          // Skip the resource based on the generation probability
+          if (!generate) {
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            return;
+          }
+          const podIRI = iri.slice(0, Math.max(0, positionContainer));
+
+          const shapeTreeIRI = `${podIRI}/${FragmentationStrategyShape.shapeTreeFileName}`;
+
+          const promises: Promise<void>[] = [];
+          // We add a triple in each file to locate the shape index if it is enable
+          if (this.tripleShapeTreeLocator === true) {
+            promises.push(FragmentationStrategyShape.generateShapeTreeLocator(quadSink, `${podIRI}/`, shapeTreeIRI, iri));
+          }
+
+          if (!this.resourcesHandled.has(resourceId)) {
+            promises.push(FragmentationStrategyShape.generateShapeIndexInformation(quadSink,
+              resourceId,
+              podIRI,
+              shapeTreeIRI,
+              directory,
+              shape,
+              positionContainerResourceNotInRoot === -1));
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            await Promise.all(promises);
+            return;
+          }
+          if (promises.length === 1) {
```

Comment:
Not sure why you have this special case here for 1-length promises.
Can't you just move the following outside of the if-statement above?
```
            this.irisHandled.add(iri);
            this.resourcesHandled.add(resourceId);
            await Promise.all(promises);
```